pkg/config/lifecycle_provider.go

```diff
@@ -0,0 +1,124 @@
+package config
+
+import (
+	"sync/atomic"
+
+	"github.com/google/go-containerregistry/pkg/authn"
+	v1 "github.com/google/go-containerregistry/pkg/v1"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	LifecycleConfigName = "lifecycle-image"
+	LifecycleConfigKey  = "image"
+)
+
+type RegistryClient interface {
+	Fetch(keychain authn.Keychain, repoName string) (v1.Image, string, error)
+}
+
+type lifecycleData struct {
+	image v1.Image
+	err   error
+}
+
+type LifecycleProvider struct {
+	RegistryClient RegistryClient
+	Keychain       authn.Keychain
+	lifecycleData  atomic.Value
+	handlers       []func()
+}
+
+func NewLifecycleProvider(lifecycleImageRef string, client RegistryClient, keychain authn.Keychain) *LifecycleProvider {
+	p := &LifecycleProvider{
+		RegistryClient: client,
+		Keychain:       keychain,
+	}
+
+	data := &lifecycleData{}
+
+	data.image, data.err = p.fetchImage(lifecycleImageRef)
+
+	p.lifecycleData.Store(data)
+	return p
+}
+
+func (l *LifecycleProvider) UpdateImage(cm *corev1.ConfigMap) {
+	data := &lifecycleData{}
+	isNewImg := true
+	defer l.callCallBacks(&isNewImg)
+	defer l.lifecycleData.Store(data)
+
+	imageRef, ok := cm.Data[LifecycleConfigKey]
+	if !ok {
+		data.err = errors.Errorf("%s config invalid", LifecycleConfigName)
+		return
+	}
+
+	newImg, err := l.fetchImage(imageRef)
+	if err != nil {
+		data.err = err
+		return
+	}
+
+	// Don't care if old image errored
+	oldImg, _ := l.GetImage()
+
+	isNewImg, err = isNewImage(oldImg, newImg)
```

Comment:
It seems that avoiding calling callbacks if the digest has not changed is actually introducing a good amount of complexity here for a rather inconsequential optimization. Could we essentially accomplish the same thing by saving the configured image in the lifecycle data and exiting early if it is the same?